redirector/go/main.go

```diff
@@ -0,0 +1,266 @@
+	mountpoint string, err error) {
+	uid := ctx.Value(fs.CtxHeaderUIDKey).(uint32)
+	// Don't let the root see anything here; we don't want a symlink
+	// loop back to this mount.
+	if uid == 0 {
+		return "", fuse.ENOENT
+	}
+
+	mountpoint = r.getCachedMountpoint(uid)
+	if mountpoint != "" {
+		return mountpoint, nil
+	}
+
+	defer func() {
+		if err != nil {
+			return
+		}
+		// Cache the entry if we didn't hit an error.
+		r.lock.Lock()
+		defer r.lock.Unlock()
+		r.mountpointCache[uid] = cacheEntry{
+			mountpoint: mountpoint,
+			time:       time.Now(),
+		}
+	}()
+
+	u, err := user.LookupId(strconv.FormatUint(uint64(uid), 10))
+	if err != nil {
+		return "", err
+	}
+
+	vols, err := gomounts.GetMountedVolumes()
+	if err != nil {
+		return "", err
+	}
+	fuseType := "fuse"
+	if runtime.GOOS == "darwin" {
+		fuseType = "kbfuse"
+	}
+	var fuseMountPoints []string
+	for _, v := range vols {
+		if v.Type != fuseType {
+			continue
+		}
+		if v.Owner != u.Uid {
+			continue
+		}
+		fuseMountPoints = append(fuseMountPoints, v.Path)
+	}
+
+	if len(fuseMountPoints) == 0 {
+		return "", fuse.ENOENT
+	}
+	if len(fuseMountPoints) == 1 {
+		return fuseMountPoints[0], nil
+	}
+
+	// If there is more than one, pick the first one alphabetically
+	// that has "keybase" in the path.
+	sort.Strings(fuseMountPoints)
+	for _, mp := range fuseMountPoints {
+		// TODO: a better regexp that will rule out keybase.staging if
+		// we're in prod mode, etc.
+		if !strings.Contains(mp, "keybase") {
+			continue
+		}
+		// Double-check that this is a real KBFS mount.
+		if _, err := os.Stat(filepath.Join(mp, ".kbfs_error")); err != nil {
+			continue
+		}
+		return mp, nil
+	}
+
+	// Give up and return the first one.
+	return fuseMountPoints[0], nil
+}
+
+func (r *root) ReadDirAll(ctx context.Context) ([]fuse.Dirent, error) {
+	_, err := r.findKBFSMount(ctx)
+	if err != nil {
+		if err == fuse.ENOENT {
+			return []fuse.Dirent{}, nil
+		}
+		return []fuse.Dirent{}, err
+	}
+
+	// TODO: show the `kbfs.error.txt" and "kbfs.nologin.txt" files if
+	// they exist?  As root, it is hard to figure out if they're
+	// there, though.
+	return []fuse.Dirent{
+		{
+			Type: fuse.DT_Link,
+			Name: "private",
+		},
+		{
+			Type: fuse.DT_Link,
+			Name: "public",
+		},
+		fuse.Dirent{
+			Type: fuse.DT_Link,
+			Name: "team",
+		},
+	}, nil
+}
+
+func (r *root) Lookup(
+	ctx context.Context, req *fuse.LookupRequest, resp *fuse.LookupResponse) (
+	n fs.Node, err error) {
+	mountpoint, err := r.findKBFSMount(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	resp.EntryValid = 0
+	switch req.Name {
+	case "private", "public", "team", ".kbfs_error", ".kbfs_metrics",
+		".kbfs_profiles", ".kbfs_reset_caches", ".kbfs_status",
+		"kbfs.error.txt", "kbfs.nologin.txt", ".kbfs_enable_auto_journals",
+		".kbfs_disable_auto_journals", ".kbfs_enable_block_prefetching",
+		".kbfs_disable_block_prefetching", ".kbfs_enable_debug_server",
+		".kbfs_disable_debug_server":
```

Comment:
Might be nice to export this set in one of our libraries.

Reply:
I didn't want to suck `libfs` and `libkbfs` into this binary, and I didn't feel like making a new one-file library for this.  But if you feel strongly about it, let me know.